src/v/redpanda/monitor_unsafe_log_flag.cc

```diff
@@ -0,0 +1,120 @@
+     * | <v23.2.1         | T          | Yes                 |
+     * | <v23.2.1         | F          | No                  |
+     * +------------------+------------+---------------------+
+     *
+     * If this is a new Redpanda cluster, then the value of the legacy flag
+     * is ignored.  If this is an upgraded cluster, then the legacy flag will
+     * control whether or not we permit control characters.
+     */
+    void handle_flag_change() {
+        if (_original_version >= flag_introduction_version) {
+            permit_unsafe_log_operation::set(false);
+        } else {
+            permit_unsafe_log_operation::set(
+              _legacy_permit_unsafe_log_operation());
+        }
+    }
+
+    cluster::cluster_version _original_version;
+    config::binding<bool> _legacy_permit_unsafe_log_operation;
+};
+
+monitor_unsafe_log_flag::monitor_unsafe_log_flag(
+  ss::abort_source& shard0_as, cluster::cluster_version original_version)
+  : _shard0_as(shard0_as)
+  , _original_version(original_version) {}
+
+ss::future<> monitor_unsafe_log_flag::maybe_log_flag_nag() {
+    auto nag_check_retry = std::chrono::seconds{300};
+    auto interval_override = std::getenv(
+      "__REDPANDA_FLAG_NAG_CHECK_INTERVAL_SEC");
+    if (interval_override != nullptr) {
+        try {
+            nag_check_retry = std::min(
+              std::chrono::seconds{nag_check_retry},
+              std::chrono::seconds{std::stoi(interval_override)});
+            vlog(
+              clusterlog.info,
+              "Overriding default flag check annoy interval to: {}s",
+              nag_check_retry.count());
+        } catch (...) {
+            vlog(
+              clusterlog.error,
+              "Invalid flag check annoy interval override '{}'",
+              interval_override);
+        }
+    }
+
+    if (
+      _original_version < flag_introduction_version
+      && permit_unsafe_log_operation::get()) {
+        vlog(
+          clusterlog.warn,
+          "You have enabled unsafe log operations.  In future versions of "
+          "Redpanda, strings containing ASCII control characters will be "
+          "rejected.  Try disabling 'legacy_permit_unsafe_log_operation'.  If "
+          "you need assistance, please contact Redpanda support");
+    }
+
+    try {
+        co_await ss::sleep_abortable(nag_check_retry, _shard0_as);
+    } catch (ss::sleep_aborted&) {
+        // Shutting down - next iteration will drop out
+    }
+}
+
+ss::future<> monitor_unsafe_log_flag::start() {
+    // Only run the nag on one shard
+    ssx::spawn_with_gate(_gate, [this] {
+        return ss::do_until(
+          [this] { return _shard0_as.abort_requested(); },
+          [this] { return maybe_log_flag_nag(); });
+    });
+
+    co_await _handle_unsafe_log_flag_change.start(_original_version);
+
+    co_return;
+}
+
+ss::future<> monitor_unsafe_log_flag::stop() {
+    co_await _handle_unsafe_log_flag_change.stop();
+    co_await _gate.close();
```

Comment:
seems like this will hang forever since the only stop condition for the background fiber is the abort source